subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/platforms/JavaPlatformResolveIntegrationTest.groovy

```diff
@@ -0,0 +1,183 @@
+            }
+        """
+    }
+
+    def "can get recommendations from a platform subproject"() {
+        def module = mavenHttpRepo.module("org", "foo", "1.1").publish()
+
+        given:
+        platformModule("""
+            constraints {
+                api "org:foo:1.1"
+            }
+        """)
+
+        buildFile << """
+            dependencies {
+                api platform(project(":platform"))
+                api "org:foo"
+            }
+        """
+        checkConfiguration("compileClasspath")
+
+        when:
+        module.pom.expectGet()
+        module.artifact.expectGet()
+
+        run ":checkDeps"
+
+        then:
+        resolve.expectGraph {
+            root(":", "org.test:test:1.9") {
+                project(":platform", "org.test:platform:1.9") {
+                    variant("api", ['org.gradle.usage': 'java-api', 'org.gradle.component.category': 'platform'])
+                    constraint("org:foo:1.1")
+                    noArtifacts()
+                }
+                edge('org:foo', 'org:foo:1.1') {
+                    byConstraint()
+                }
+            }
+        }
+    }
+
+    def "can get different recommendations from a platform runtime subproject"() {
+        def module1 = mavenHttpRepo.module("org", "foo", "1.1").publish()
+        def module2 = mavenHttpRepo.module("org", "foo", "1.2").publish()
+
+        given:
+        platformModule("""
+            constraints {
+                api "org:foo:1.1"
+                runtime "org:foo:1.2"
+            }
+        """)
+
+        buildFile << """
+            dependencies {
+                api platform(project(":platform"))
+                api "org:foo"
+            }
+        """
+        checkConfiguration("runtimeClasspath")
+
+        when:
+        module2.pom.expectGet()
+        module2.artifact.expectGet()
+
+        run ":checkDeps"
+
+        then:
+        resolve.expectGraph {
+            root(":", "org.test:test:1.9") {
+                project(":platform", "org.test:platform:1.9") {
+                    variant("runtime", ['org.gradle.usage': 'java-runtime', 'org.gradle.component.category': 'platform'])
+                    constraint("org:foo:1.1", "org:foo:1.2") // this is just because of the dumb test
+                    constraint("org:foo:1.2")
+                    noArtifacts()
+                }
+                edge('org:foo', 'org:foo:1.2') {
+                    configuration = "runtime"
+                    byConstraint()
+                }
+            }
+        }
+    }
+
+    def "reasonable behavior when using a regular project dependency instead of a platform dependency"() {
+        def module = mavenHttpRepo.module("org", "foo", "1.1").publish()
+
+        given:
+        platformModule("""
+            constraints {
+                api "org:foo:1.1"
+            }
+        """)
+
+        buildFile << """
+            dependencies {
+                api project(":platform")
```

Comment:
Does that mean that consuming a Gradle platform without `platform` works just as well? I am guessing the behavior would be different for Maven. Right?
Or is it only the case because you are using a project dependency?

Reply:
It does work thanks to disambiguation rules. However we might want to make this an error. Not sure.